File: ios/Classes/Base/RtcChannel.swift
//
//  RtcChannel.swift
//  RCTAgora
//
//  Created by LXH on 2020/4/13.
//  Copyright (c) 2020 Syan. All rights reserved.
//

import AgoraRtcKit
import Foundation

protocol RtcChannelInterface:
        RtcChannelAudioInterface,
        RtcChannelVideoInterface,
        RtcChannelVoicePositionInterface,
        RtcChannelPublishStreamInterface,
        RtcChannelMediaRelayInterface,
        RtcChannelDualStreamInterface,
        RtcChannelFallbackInterface,
        RtcChannelMediaMetadataInterface,
        RtcChannelEncryptionInterface,
        RtcChannelInjectStreamInterface,
        RtcChannelStreamMessageInterface {
    func create(_ params: NSDictionary, _ callback: Callback)

    func destroy(_ params: NSDictionary, _ callback: Callback)

    func setClientRole(_ params: NSDictionary, _ callback: Callback)

    func joinChannel(_ params: NSDictionary, _ callback: Callback)

    func joinChannelWithUserAccount(_ params: NSDictionary, _ callback: Callback)

    func leaveChannel(_ params: NSDictionary, _ callback: Callback)

    func renewToken(_ params: NSDictionary, _ callback: Callback)

    func getConnectionState(_ params: NSDictionary, _ callback: Callback)

    @available(*, deprecated)
    func publish(_ params: NSDictionary, _ callback: Callback)

    @available(*, deprecated)
    func unpublish(_ params: NSDictionary, _ callback: Callback)

    func getCallId(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelAudioInterface {
    func adjustUserPlaybackSignalVolume(_ params: NSDictionary, _ callback: Callback)

    func muteLocalAudioStream(_ params: NSDictionary, _ callback: Callback)

    func muteRemoteAudioStream(_ params: NSDictionary, _ callback: Callback)

    func muteAllRemoteAudioStreams(_ params: NSDictionary, _ callback: Callback)

    @available(*, deprecated)
    func setDefaultMuteAllRemoteAudioStreams(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelVideoInterface {
    func muteLocalVideoStream(_ params: NSDictionary, _ callback: Callback)

    func muteRemoteVideoStream(_ params: NSDictionary, _ callback: Callback)

    func muteAllRemoteVideoStreams(_ params: NSDictionary, _ callback: Callback)

    @available(*, deprecated)
    func setDefaultMuteAllRemoteVideoStreams(_ params: NSDictionary, _ callback: Callback)

    func enableRemoteSuperResolution(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelVoicePositionInterface {
    func setRemoteVoicePosition(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelPublishStreamInterface {
    func setLiveTranscoding(_ params: NSDictionary, _ callback: Callback)

    func addPublishStreamUrl(_ params: NSDictionary, _ callback: Callback)

    func removePublishStreamUrl(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelMediaRelayInterface {
    func startChannelMediaRelay(_ params: NSDictionary, _ callback: Callback)

    func updateChannelMediaRelay(_ params: NSDictionary, _ callback: Callback)

    func pauseAllChannelMediaRelay(_ params: NSDictionary, _ callback: Callback)

    func resumeAllChannelMediaRelay(_ params: NSDictionary, _ callback: Callback)

    func stopChannelMediaRelay(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelDualStreamInterface {
    func setRemoteVideoStreamType(_ params: NSDictionary, _ callback: Callback)

    func setRemoteDefaultVideoStreamType(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelFallbackInterface {
    func setRemoteUserPriority(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelMediaMetadataInterface {
    func registerMediaMetadataObserver(_ params: NSDictionary, _ callback: Callback)

    func unregisterMediaMetadataObserver(_ params: NSDictionary, _ callback: Callback)

    func setMaxMetadataSize(_ params: NSDictionary, _ callback: Callback)

    func sendMetadata(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelEncryptionInterface {
    @available(*, deprecated)
    func setEncryptionSecret(_ params: NSDictionary, _ callback: Callback)

    @available(*, deprecated)
    func setEncryptionMode(_ params: NSDictionary, _ callback: Callback)

    func enableEncryption(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelInjectStreamInterface {
    func addInjectStreamUrl(_ params: NSDictionary, _ callback: Callback)

    func removeInjectStreamUrl(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelStreamMessageInterface {
    func createDataStream(_ params: NSDictionary, _ callback: Callback)

    func sendStreamMessage(_ params: NSDictionary, _ callback: Callback)
}

@objc
class RtcChannelManager: NSObject, RtcChannelInterface {
    private var emitter: (_ methodName: String, _ data: [String: Any?]?) -> Void
    private var rtcChannelMap = [String: AgoraRtcChannel]()
    private var rtcChannelDelegateMap = [String: RtcChannelEventHandler]()
    private var mediaObserverMap = [String: MediaObserver]()

    init(_ emitter: @escaping (_ methodName: String, _ data: [String: Any?]?) -> Void) {
        self.emitter = emitter
    }

    func Release() {
        rtcChannelMap.forEach {
            $1.destroy()
        }
        rtcChannelMap.removeAll()
        rtcChannelDelegateMap.removeAll()
        mediaObserverMap.removeAll()
    }

    subscript(channelId: String) -> AgoraRtcChannel? {
        return rtcChannelMap[channelId]
    }

    @objc func create(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(params["engine"] as? AgoraRtcEngineKit) { [weak self] in
            if let rtcChannel = $0.createRtcChannel(params["channelId"] as! String) {
                let delegate = RtcChannelEventHandler { [weak self] in
                    self?.emitter($0, $1)
                }
                rtcChannel.setRtcChannelDelegate(delegate)
                self?.rtcChannelMap[rtcChannel.getId()!] = rtcChannel
                self?.rtcChannelDelegateMap[rtcChannel.getId()!] = delegate
            }
            return nil
        }
    }

    @objc func destroy(_ params: NSDictionary, _ callback: Callback) {
        callback.code(rtcChannelMap.removeValue(forKey: params["channelId"] as! String)?.destroy())
    }

    @objc func setClientRole(_ params: NSDictionary, _ callback: Callback) {
        let role = AgoraClientRole(rawValue: (params["role"] as! NSNumber).intValue)!
        if let options = params["options"] as? [String: Any] {
            callback.code(self[params["channelId"] as! String]?.setClientRole(role, options: mapToClientRoleOptions(options)))
            return
        }
        callback.code(self[params["channelId"] as! String]?.setClientRole(role))
    }

    @objc func joinChannel(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.join(byToken: params["token"] as? String, info: params["optionalInfo"] as? String, uid: (params["optionalUid"] as! NSNumber).uintValue, options: mapToChannelMediaOptions(params["options"] as! [String: Any])))
    }

    @objc func joinChannelWithUserAccount(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.join(byUserAccount: params["userAccount"] as! String, token: params["token"] as? String, options: mapToChannelMediaOptions(params["options"] as! [String: Any])))
    }

    @objc func leaveChannel(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.leave())
    }

    @objc func renewToken(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.renewToken(params["token"] as! String))
    }

    @objc func getConnectionState(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(self[params["channelId"] as! String]) {
            $0.getConnectionState().rawValue
        }
    }

    @objc func publish(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.publish())
    }

    @objc func unpublish(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.unpublish())
    }

    @objc func getCallId(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(self[params["channelId"] as! String]) {
            $0.getCallId()
        }
    }

    @objc func adjustUserPlaybackSignalVolume(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.adjustUserPlaybackSignalVolume((params["uid"] as! NSNumber).uintValue, volume: (params["volume"] as! NSNumber).int32Value))
    }

    @objc func muteRemoteAudioStream(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteRemoteAudioStream((params["uid"] as! NSNumber).uintValue, mute: params["muted"] as! Bool))
    }

    @objc func muteAllRemoteAudioStreams(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteAllRemoteAudioStreams(params["muted"] as! Bool))
    }

    @objc func setDefaultMuteAllRemoteAudioStreams(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setDefaultMuteAllRemoteAudioStreams(params["muted"] as! Bool))
    }

    @objc func muteRemoteVideoStream(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteRemoteVideoStream((params["uid"] as! NSNumber).uintValue, mute: params["muted"] as! Bool))
    }

    @objc func muteAllRemoteVideoStreams(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteAllRemoteVideoStreams(params["muted"] as! Bool))
    }

    @objc func setDefaultMuteAllRemoteVideoStreams(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setDefaultMuteAllRemoteVideoStreams(params["muted"] as! Bool))
    }

    @objc func setRemoteVoicePosition(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setRemoteVoicePosition((params["uid"] as! NSNumber).uintValue, pan: (params["pan"] as! NSNumber).doubleValue, gain: (params["gain"] as! NSNumber).doubleValue))
    }

    @objc func setLiveTranscoding(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setLiveTranscoding(mapToLiveTranscoding(params["transcoding"] as! [String: Any])))
    }

    @objc func addPublishStreamUrl(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.addPublishStreamUrl(params["url"] as! String, transcodingEnabled: params["transcodingEnabled"] as! Bool))
    }

    @objc func removePublishStreamUrl(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.removePublishStreamUrl(params["url"] as! String))
    }

    @objc func startChannelMediaRelay(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.startMediaRelay(mapToChannelMediaRelayConfiguration(params["channelMediaRelayConfiguration"] as! [String: Any])))
    }

    @objc func updateChannelMediaRelay(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.updateMediaRelay(mapToChannelMediaRelayConfiguration(params["channelMediaRelayConfiguration"] as! [String: Any])))
    }

    @objc func stopChannelMediaRelay(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.stopMediaRelay())
    }

    @objc func setRemoteVideoStreamType(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setRemoteVideoStream((params["uid"] as! NSNumber).uintValue, type: AgoraVideoStreamType(rawValue: (params["streamType"] as! NSNumber).intValue)!))
    }

    @objc func setRemoteDefaultVideoStreamType(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setRemoteDefaultVideoStreamType(AgoraVideoStreamType(rawValue: (params["streamType"] as! NSNumber).intValue)!))
    }

    @objc func setRemoteUserPriority(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setRemoteUserPriority((params["uid"] as! NSNumber).uintValue, type: AgoraUserPriority(rawValue: (params["userPriority"] as! NSNumber).intValue)!))
    }

    @objc func registerMediaMetadataObserver(_ params: NSDictionary, _ callback: Callback) {
        let channelId = params["channelId"] as! String
        let mediaObserver = MediaObserver { [weak self] in
            if var data = $0 {
                data["channelId"] = channelId
                self?.emitter(RtcEngineEvents.MetadataReceived, data)
            }
        }
        callback.resolve(self[channelId]) {
            if $0.setMediaMetadataDelegate(mediaObserver, with: .video) {
                mediaObserverMap[channelId] = mediaObserver
            }
            return nil
        }
    }

    @objc func unregisterMediaMetadataObserver(_ params: NSDictionary, _ callback: Callback) {
        let channelId = params["channelId"] as! String
        callback.resolve(self[channelId]) {
            if $0.setMediaMetadataDelegate(nil, with: .video) {
                mediaObserverMap.removeValue(forKey: channelId)
            }
            return nil
        }
    }

    @objc func setMaxMetadataSize(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(mediaObserverMap[params["channelId"] as! String]) {
            $0.setMaxMetadataSize((params["size"] as! NSNumber).intValue)
        }
    }

    @objc func sendMetadata(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(mediaObserverMap[params["channelId"] as! String]) {
            $0.addMetadata(params["metadata"] as! String)
        }
    }

    @objc func setEncryptionSecret(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setEncryptionSecret(params["secret"] as? String))
    }

    @objc func setEncryptionMode(_ params: NSDictionary, _ callback: Callback) {
        var encryptionMode = ""
        switch (params["encryptionMode"] as! NSNumber).intValue {
        case AgoraEncryptionMode.AES128XTS.rawValue:
            encryptionMode = "aes-128-xts"
        case AgoraEncryptionMode.AES128ECB.rawValue:
            encryptionMode = "aes-128-ecb"
        case AgoraEncryptionMode.AES256XTS.rawValue:
            encryptionMode = "aes-256-xts"
        default: encryptionMode = ""
        }
        callback.code(self[params["channelId"] as! String]?.setEncryptionMode(encryptionMode))
    }

    @objc func enableEncryption(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.enableEncryption(params["enabled"] as! Bool, encryptionConfig: mapToEncryptionConfig(params["config"] as! [String: Any])))
    }

    @objc func addInjectStreamUrl(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.addInjectStreamUrl(params["url"] as! String, config: mapToLiveInjectStreamConfig(params["config"] as! [String: Any])))
    }

    @objc func removeInjectStreamUrl(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.removeInjectStreamUrl(params["url"] as! String))
    }

    @objc func createDataStream(_ params: NSDictionary, _ callback: Callback) {
        let channel = self[params["channelId"] as! String]
        var streamId = 0
        if let config = params["config"] as? [String: Any] {
            callback.code(channel?.createDataStream(&streamId, config: mapToDataStreamConfig(config))) { _ in
                streamId
            }
            return
        }
        callback.code(channel?.createDataStream(&streamId, reliable: params["reliable"] as! Bool, ordered: params["ordered"] as! Bool)) { _ in
            streamId
        }
    }

    @objc func sendStreamMessage(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.sendStreamMessage((params["streamId"] as! NSNumber).intValue, data: (params["message"] as! String).data(using: .utf8)!))
    }

    @objc func enableRemoteSuperResolution(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.enableRemoteSuperResolution((params["uid"] as! NSNumber).uintValue, enabled: params["enabled"] as! Bool))
    }

    @objc func muteLocalAudioStream(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteLocalAudioStream(params["muted"] as! Bool))
    }

    @objc func muteLocalVideoStream(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteLocalVideoStream(params["muted"] as! Bool))
    }

    @objc func pauseAllChannelMediaRelay(_ params: NSDictionary, _ callback: Callback) {
        callback.code(-Int32(AgoraErrorCode.notSupported.rawValue))
//        callback.code(self[params["channelId"] as! String]?.pauseAllChannelMediaRelay())
    }

    @objc func resumeAllChannelMediaRelay(_ params: NSDictionary, _ callback: Callback) {
        callback.code(-Int32(AgoraErrorCode.notSupported.rawValue))
//        callback.code(self[params["channelId"] as! String]?.resumeAllChannelMediaRelay())
    }
}

File: RtcChannel.kt
package io.agora.rtc.base

import io.agora.rtc.Constants
import io.agora.rtc.IMetadataObserver
import io.agora.rtc.RtcChannel
import io.agora.rtc.RtcEngine
import io.agora.rtc.internal.EncryptionConfig
import java.util.*

class IRtcChannel {
    interface RtcChannelInterface : RtcAudioInterface, RtcVideoInterface, RtcVoicePositionInterface,
            RtcPublishStreamInterface, RtcMediaRelayInterface, RtcDualStreamInterface,
            RtcFallbackInterface, RtcMediaMetadataInterface, RtcEncryptionInterface,
            RtcInjectStreamInterface, RtcStreamMessageInterface {
        fun create(params: Map<String, *>, callback: Callback)

        fun destroy(params: Map<String, *>, callback: Callback)

        fun setClientRole(params: Map<String, *>, callback: Callback)

        fun joinChannel(params: Map<String, *>, callback: Callback)

        fun joinChannelWithUserAccount(params: Map<String, *>, callback: Callback)

        fun leaveChannel(params: Map<String, *>, callback: Callback)

        fun renewToken(params: Map<String, *>, callback: Callback)

        fun getConnectionState(params: Map<String, *>, callback: Callback)

        fun publish(params: Map<String, *>, callback: Callback)

        fun unpublish(params: Map<String, *>, callback: Callback)

        fun getCallId(params: Map<String, *>, callback: Callback)
    }

    interface RtcAudioInterface {
        fun adjustUserPlaybackSignalVolume(params: Map<String, *>, callback: Callback)

        fun muteRemoteAudioStream(params: Map<String, *>, callback: Callback)

        fun muteAllRemoteAudioStreams(params: Map<String, *>, callback: Callback)

        fun setDefaultMuteAllRemoteAudioStreams(params: Map<String, *>, callback: Callback)
    }

    interface RtcVideoInterface {
        fun muteRemoteVideoStream(params: Map<String, *>, callback: Callback)

        fun muteAllRemoteVideoStreams(params: Map<String, *>, callback: Callback)

        fun setDefaultMuteAllRemoteVideoStreams(params: Map<String, *>, callback: Callback)
    }

    interface RtcVoicePositionInterface {
        fun setRemoteVoicePosition(params: Map<String, *>, callback: Callback)
    }

    interface RtcPublishStreamInterface {
        fun setLiveTranscoding(params: Map<String, *>, callback: Callback)

        fun addPublishStreamUrl(params: Map<String, *>, callback: Callback)

        fun removePublishStreamUrl(params: Map<String, *>, callback: Callback)
    }

    interface RtcMediaRelayInterface {
        fun startChannelMediaRelay(params: Map<String, *>, callback: Callback)

        fun updateChannelMediaRelay(params: Map<String, *>, callback: Callback)

        fun stopChannelMediaRelay(params: Map<String, *>, callback: Callback)
    }

    interface RtcDualStreamInterface {
        fun setRemoteVideoStreamType(params: Map<String, *>, callback: Callback)

        fun setRemoteDefaultVideoStreamType(params: Map<String, *>, callback: Callback)
    }

    interface RtcFallbackInterface {
        fun setRemoteUserPriority(params: Map<String, *>, callback: Callback)
    }

    interface RtcMediaMetadataInterface {
        fun registerMediaMetadataObserver(params: Map<String, *>, callback: Callback)

        fun unregisterMediaMetadataObserver(params: Map<String, *>, callback: Callback)

        fun setMaxMetadataSize(params: Map<String, *>, callback: Callback)

        fun sendMetadata(params: Map<String, *>, callback: Callback)
    }

    interface RtcEncryptionInterface {
        fun setEncryptionSecret(params: Map<String, *>, callback: Callback)

        fun setEncryptionMode(params: Map<String, *>, callback: Callback)

        fun enableEncryption(params: Map<String, *>, callback: Callback)
    }

    interface RtcInjectStreamInterface {
        fun addInjectStreamUrl(params: Map<String, *>, callback: Callback)

        fun removeInjectStreamUrl(params: Map<String, *>, callback: Callback)
    }

    interface RtcStreamMessageInterface {
        fun createDataStream(params: Map<String, *>, callback: Callback)

        fun sendStreamMessage(params: Map<String, *>, callback: Callback)
    }
}

class RtcChannelManager(
        private val emit: (methodName: String, data: Map<String, Any?>?) -> Unit
) : IRtcChannel.RtcChannelInterface {
    private val rtcChannelMap = Collections.synchronizedMap(mutableMapOf<String, RtcChannel>())
    private val mediaObserverMap = Collections.synchronizedMap(mutableMapOf<String, MediaObserver>())

    fun release() {
        rtcChannelMap.forEach { it.value.destroy() }
        rtcChannelMap.clear()
        mediaObserverMap.clear()
    }

    operator fun get(channelId: String): RtcChannel? {
        return rtcChannelMap[channelId]
    }

    override fun create(params: Map<String, *>, callback: Callback) {
        callback.resolve(params["engine"] as RtcEngine) { e ->
            e.createRtcChannel(params["channelId"] as String)?.let {
                it.setRtcChannelEventHandler(RtcChannelEventHandler { methodName, data -> emit(methodName, data) })
                rtcChannelMap[it.channelId()] = it
            }
            Unit
        }
    }

    override fun destroy(params: Map<String, *>, callback: Callback) {
        var code: Int? = -Constants.ERR_NOT_INITIALIZED
        this[params["channelId"] as String]?.let {
            code = rtcChannelMap.remove(it.channelId())?.destroy()
        }
        callback.code(code)
    }

    override fun setClientRole(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setClientRole((params["role"] as Number).toInt()))
    }

    override fun joinChannel(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.joinChannel(params["token"] as? String, params["optionalInfo"] as? String, (params["optionalUid"] as Number).toInt(), mapToChannelMediaOptions(params["options"] as Map<*, *>)))
    }

    override fun joinChannelWithUserAccount(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.joinChannelWithUserAccount(params["token"] as? String, params["userAccount"] as String, mapToChannelMediaOptions(params["options"] as Map<*, *>)))
    }

    override fun leaveChannel(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.leaveChannel())
    }

    override fun renewToken(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.renewToken(params["token"] as String))
    }

    override fun getConnectionState(params: Map<String, *>, callback: Callback) {
        callback.resolve(this[params["channelId"] as String]) { it.connectionState }
    }

    override fun publish(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.publish())
    }

    override fun unpublish(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.unpublish())
    }

    override fun getCallId(params: Map<String, *>, callback: Callback) {
        callback.resolve(this[params["channelId"] as String]) { it.callId }
    }

    override fun adjustUserPlaybackSignalVolume(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.adjustUserPlaybackSignalVolume((params["uid"] as Number).toInt(), (params["volume"] as Number).toInt()))
    }

    override fun muteRemoteAudioStream(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.muteRemoteAudioStream((params["uid"] as Number).toInt(), params["muted"] as Boolean))
    }

    override fun muteAllRemoteAudioStreams(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.muteAllRemoteAudioStreams(params["muted"] as Boolean))
    }

    override fun setDefaultMuteAllRemoteAudioStreams(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setDefaultMuteAllRemoteAudioStreams(params["muted"] as Boolean))
    }

    override fun muteRemoteVideoStream(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.muteRemoteVideoStream((params["uid"] as Number).toInt(), params["muted"] as Boolean))
    }

    override fun muteAllRemoteVideoStreams(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.muteAllRemoteVideoStreams(params["muted"] as Boolean))
    }

    override fun setDefaultMuteAllRemoteVideoStreams(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setDefaultMuteAllRemoteVideoStreams(params["muted"] as Boolean))
    }

    override fun setRemoteVoicePosition(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setRemoteVoicePosition((params["uid"] as Number).toInt(), (params["pan"] as Number).toDouble(), (params["gain"] as Number).toDouble()))
    }

    override fun setLiveTranscoding(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setLiveTranscoding(mapToLiveTranscoding(params["transcoding"] as Map<*, *>)))
    }

    override fun addPublishStreamUrl(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.addPublishStreamUrl(params["url"] as String, params["transcodingEnabled"] as Boolean))
    }

    override fun removePublishStreamUrl(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.removePublishStreamUrl(params["url"] as String))
    }

    override fun startChannelMediaRelay(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.startChannelMediaRelay(mapToChannelMediaRelayConfiguration(params["channelMediaRelayConfiguration"] as Map<*, *>)))
    }

    override fun updateChannelMediaRelay(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.updateChannelMediaRelay(mapToChannelMediaRelayConfiguration(params["channelMediaRelayConfiguration"] as Map<*, *>)))
    }

    override fun stopChannelMediaRelay(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.stopChannelMediaRelay())
    }

    override fun setRemoteVideoStreamType(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setRemoteVideoStreamType((params["uid"] as Number).toInt(), (params["streamType"] as Number).toInt()))
    }

    override fun setRemoteDefaultVideoStreamType(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setRemoteDefaultVideoStreamType((params["streamType"] as Number).toInt()))
    }

    override fun setRemoteUserPriority(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setRemoteUserPriority((params["uid"] as Number).toInt(), (params["userPriority"] as Number).toInt()))
    }

    override fun registerMediaMetadataObserver(params: Map<String, *>, callback: Callback) {
        var code = -Constants.ERR_NOT_INITIALIZED
        this[params["channelId"] as String]?.let {
            val mediaObserver = MediaObserver { data ->
                emit(RtcChannelEvents.MetadataReceived, data?.toMutableMap()?.apply { put("channelId", it.channelId()) })
            }
            code = it.registerMediaMetadataObserver(mediaObserver, IMetadataObserver.VIDEO_METADATA)
            if (code == 0) mediaObserverMap[it.channelId()] = mediaObserver
        }
        callback.code(code)
    }

    override fun unregisterMediaMetadataObserver(params: Map<String, *>, callback: Callback) {
        var code = -Constants.ERR_NOT_INITIALIZED
        this[params["channelId"] as String]?.let {
            code = it.registerMediaMetadataObserver(null, IMetadataObserver.VIDEO_METADATA)
            if (code == 0) mediaObserverMap.remove(it.channelId())
        }
        callback.code(code)
    }

    override fun setMaxMetadataSize(params: Map<String, *>, callback: Callback) {
        callback.resolve(mediaObserverMap[params["channelId"] as String]) {
            it.maxMetadataSize = (params["size"] as Number).toInt()
            Unit
        }
    }

    override fun sendMetadata(params: Map<String, *>, callback: Callback) {
        callback.resolve(mediaObserverMap[params["channelId"] as String]) {
            it.addMetadata(params["metadata"] as String)
            Unit
        }
    }

    override fun setEncryptionSecret(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setEncryptionSecret(params["secret"] as String))
    }

    override fun setEncryptionMode(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setEncryptionMode(when ((params["encryptionMode"] as Number).toInt()) {
            EncryptionConfig.EncryptionMode.AES_128_XTS.value -> "aes-128-xts"
            EncryptionConfig.EncryptionMode.AES_128_ECB.value -> "aes-128-ecb"
            EncryptionConfig.EncryptionMode.AES_256_XTS.value -> "aes-256-xts"
            else -> ""
        }))
    }

    override fun enableEncryption(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.enableEncryption(params["enabled"] as Boolean, mapToEncryptionConfig(params["config"] as Map<*, *>)))
    }

    override fun addInjectStreamUrl(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.addInjectStreamUrl(params["url"] as String, mapToLiveInjectStreamConfig(params["config"] as Map<*, *>)))
    }

    override fun removeInjectStreamUrl(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.removeInjectStreamUrl(params["url"] as String))
    }

    override fun createDataStream(params: Map<String, *>, callback: Callback) {
        var code = -Constants.ERR_NOT_INITIALIZED
        this[params["channelId"] as String]?.let {
            code = it.createDataStream(params["reliable"] as Boolean, params["ordered"] as Boolean)
        }
        callback.code(code) { it }
    }

    override fun sendStreamMessage(params: Map<String, *>, callback: Callback) {
        var code = -Constants.ERR_NOT_INITIALIZED
        this[params["channelId"] as String]?.let {
            code = it.sendStreamMessage((params["streamId"] as Number).toInt(), (params["message"] as String).toByteArray())
        }
        callback.code(code)
    }
}
